rfcs/45-stageless.md

```diff
@@ -2,488 +2,474 @@
     }
 }
 ```
 
-Just like systems, system sets can be given run criteria and be part of other sets themselves, allowing users to describe properties hierarchically.
+### Configuring run conditions (previously run criteria)
 
-#### System ordering
+While ordering constraints determine *when* a system runs, **run conditions** determine *if* it runs at all. Run conditions consist of systems that have read-only data access and return `bool`.
+A system or set can have any number of conditions, and it will only run if all of them return `true`. 
+If any condition returns `false`, the system (or systems under the set) will be skipped.
 
-The most important way we can configure systems is by telling the scheduler *when* they should be run.
-The ordering of systems is always defined relative to other systems: either directly or by checking the systems that belong to a set.
-
-Applying an ordering constraint between sets causes ordering constraints to be created between all individual members of those sets.
-If an ordering is defined relative to an empty set, it will have no effect, and a warning will be emitted.
-This relatively gentle failure mode is important to ensure that plugins can order their systems with relatively strong assumptions that the default system sets exist, but continue to (mostly) work if those systems or sets are not present.
-
-In addition to the `.before` and `.after` methods, you can use **system chains** to create very simple linear dependencies between the successive members of an array of systems.
-(Note to readers: this is not the same as "system chaining" in Bevy 0.7 and earlier: that concept has been renamed to "system piping".)
-
-```rust
-fn main(){
-   App::new()
-   // These systems are connected using a string of ordering constraints
-
-   .add_systems((compute_attack, 
-                 compute_defense,
-                 check_for_crits,
-                 compute_damage,
-                 deal_damage,
-                 check_for_death)
-                 .chain()
-                 // We can configure all systems in the chain at once
-                 .in_set(GameSet::Combat)
-   )
-   .run()
-}
-```
-
-#### Run criteria
-
-While ordering constraints determine *when* a system will run, **run criteria** will determine *if* it will run at all.
-Each system can have any number of run criteria, which read data from the world to control whether or not that system runs.
-If (and only if) all of its run criteria return `true`, the system will run.
-If any of the run criteria are `false`, the system will be skipped.
-Systems that are skipped are considered completed for the purposes of ordering constraints.
-Run criteria are not systems, but can be created from systems which can read (but not write) data from the `World` and returns a boolean value.
-
-You can specify run criteria in several different ways:
+Conditions are each evaluated *at most once* during a single pass of a schedule.

```

Comment:
My understanding that this is just for conditions applied to a single set/label. If you add a condition twice it would be considered as 2 different conditions, similar to how systems added multiple times to the schedule are really different systems.

Reply:
If I'm reading the draft PR correctly, you could even add the same run criteria to the set a system is under and separately to the system itself (e.g. if some resource exists) and the run condition will be potentially evaluated twice. 

The purpose of this sentence is probably clarifying that a run condition is only evaluated once per a set, rather than once for every system in the set. This was how it was going to be done in a previous versions of stageless.

Reply:
That seems reasonable. In fact, @maniwani explained the new approach in [this comment]. I think that either this is an outdated piece of information, or should be clarified.


[this comment]: https://github.com/bevyengine/rfcs/pull/45#discussion_r864751917

Reply:
The "at most once" thing was about the hierarchy. A set can have conditions and if one of them returns `false`, then its systems, sub-sets, etc. are skipped along with their conditions.

So during each pass of a schedule, a condition system will run only if the thing it's attached to isn't skipped earlier.